src/mainboard/starfive/visionfive2/bt0/src/init.rs

```diff
@@ -0,0 +1,677 @@
+pub const RSTGEN_SOFT_ASSERT3: u32 = RSTGEN_BASE + 0xC;
+
+pub const RSTGEN_SOFT_STATUS0: u32 = RSTGEN_BASE + 0x10;
+pub const RSTGEN_SOFT_STATUS1: u32 = RSTGEN_BASE + 0x14;
+pub const RSTGEN_SOFT_STATUS2: u32 = RSTGEN_BASE + 0x18;
+pub const RSTGEN_SOFT_STATUS3: u32 = RSTGEN_BASE + 0x1C;
+
+pub fn clear_rstgen_rstn_usbnoc_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 6);
+    v |= 0 << 6;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 6;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_hifi4noc_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 2);
+    v |= 0 << 2;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 2;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_x2c_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 9);
+    v |= 0 << 9;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 9;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn assert_rstgen_rstn_x2c_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 9);
+    v |= 1 << 9;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 9;
+        v &= 0x1;
+        if !v != 0x0 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_msi_apb_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT3);
+    v &= !(0x1 << 14);
+    v |= 0 << 14;
+    write_32(RSTGEN_SOFT_ASSERT3, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS3) >> 14;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn assert_rstgen_rstn_ddrphy_ahb() {
+    let v = read_32(RSTGEN_SOFT_ASSERT1) & !(1 << 30);
+    write_32(RSTGEN_SOFT_ASSERT1, v | 1 << 30);
+    loop {
+        if !((read_32(RSTGEN_SOFT_STATUS1) >> 30) & 1) != 0 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_dspx2c_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 14);
+    v |= 0 << 14;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 14;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_dma1p_axi_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 8);
+    v |= 0 << 8;
+    write_32(RSTGEN_SOFT_ASSERT1, v);
+    loop {
+        let mut v = read_32(RSTGEN_SOFT_STATUS1) >> 8;
+        v &= 0x1;
+        if !v != 0x1 {
+            break;
+        }
+    }
+}
+
+pub fn clear_rstgen_rstn_gmac_ahb_() {
+    let mut v = read_32(RSTGEN_SOFT_ASSERT1);
+    v &= !(0x1 << 28);
+    v |= 0 << 28;
```

Comment:
Isn't this a no-op?

Reply:
Yes, and you need not care about this... this whole chunk here is copied from the previous SoC and just here for reference... may not be applicable at all. This is all a draft, and what I am currently on is figuring out issues with print macros...

Reply:
If it's a no-op, and serves no purpose, wouldn't it be better to just remove it?

> This is all a draft, and what I am currently on is figuring out issues with print macros...

Understood, if you want debugging/testing help, let me know.

Reply:
See the latest commits I pushed.
For reference, this happens:
https://twitter.com/OrangeCMS/status/1632208849254981633